doc/dev/kdb-operations.md

```diff
@@ -133,53 +133,47 @@ The purpose of the `set` operation is to write data from a `KDB` instance into b
 
 Properties of `kdbSet()`:
 
+<!-- TODO: what about spec/copy? Should that really not be reflected? With the steps below the changes would show up... -->
+
 - When calling `kdbSet (kdb, ks, parentKey)` the contents (key names, values and metadata) of `ks` _will not be modified_.
 - _All keys_ in `ks` that are below `parentKey` will be persisted in the KDB, when a `kdbSet (kdb, ks, parentKey)` call returns successfully.
   Additionally, any key in `ks` that shares a backend with another key which is below `parentKey` will also be persisted.
-- Calling `kdbSet` results in an error, if `kdbGet` wasn't called on this `KDB` instance at least once.
-
-TODO: notifications? (just call at the end?)
-
-The basic flow of this operation is:
+- Calling `kdbSet` may result in an error, if `kdbGet` wasn't called on this `KDB` instance with the same `parentKey` at least once.
 
-<!-- TODO: check optimizations (OPT), remove unless clearly useful -->
+The flow of this operation is:
 
 1. Determine the backends needed to write all keys below `parentKey`.
-2. Check that all backends are initialized (i.e. `kdbGet()` was called).
-3. Run the `spec` plugin on `ks` (to add metakeys for new keys).
-   <!-- TODO: Running spec at this point makes technical sense, but isn't very logical otherwise. -->
-   <!-- TODO: Could we merge the deep-copy, split and read-only/unchanged check steps? -->
+2. Check that all backends are opened and initialized (i.e. `kdbGet()` was called).
+3. Run the `spec/copy` hook on `ks` (to add metakeys to newly created keys).
 4. Deep-Copy `ks` (below `parentKey`) into a new KeySet `set_ks`
 5. Split `set_ks` into individual backends
 6. Determine which backends contain changed data.
    Any backend that contains a key that needs sync (via `KEY_FLAG_SYNC`) could contain changed data.
    From now on ignore all backends that have not changed.
    From now on also ignore all backends that were initialized as read-only.
    Issue a warning, if a change was detected (via `KEY_FLAG_SYNC`) in a read-only backend.
+   > **Note**: Steps 4-6 might be combined into a single procedure that deep-copies only keys from changed backends into separate KeySets per backend
 7. Run the `resolver` and `prestorage` on all backends (abort immediately on error and go to e).
-<!-- TODO: can this merge+split be avoided?
-    do we really need to call spec for this?
-    Could we use metaspec:/ namespace and just remove all those keys?
- -->
+<!-- TODO: instead of this merge, spec/remove, split sequence, maybe we could add a metaspec:/ namespace and simply remove all those metakeys? -->
```

Comment:
Adding a `metaspec:/` namespace is pretty clean, but it is also a lot of effort and has some drawbacks (e.g. no more direct `keyCopyMeta`).

We can't call `spec/remove` separately for all mountpoints, because that would fail for e.g. `spec:/foo` and `user:/foo/bar` as mountpoints. The spec from `spec:/foo` wouldn't apply to keys below `user:/foo/bar`.

However, if `spec/remove` only ever modifies existing keys (which should be possible), we only need the merge and can simply not do the second split. We would just create a `KeySet` that combines alle the results from Step 7 and run `spec/remove` on that. Because this new `KeySet` and the ones from Step 7 share the same `Key`s, we could just use the ones from Step 7 after `spec/remove`.

Reply:
I don't get the comment. Is there a question? metaspec is definitely out of scope for new-backend so it also shouldn't be mentioned in the PR.

Reply:
This is basically ideas about optimizing the operational sequence of `kdbSet`. Currently we need to split `ks` into the backends, do some work, then merge everything back into on `KeySet`, call `spec/remove` and split it back into the backends. With `metaspec:/` we wouldn't need the "merge, `spec/remove`, split" steps. Instead we could remove all `metaspec:/` keys directly in `kdbSet` and not have the `spec/remove` hook at all. Even without `metaspec:/` I think it may be possible (with a careful implementation of `spec/remove`) to only do "merge, `spec/remove`" and but not the "split" step.

I will just add a note that these are possible future optimizations. The only actual question is whether `metaspec:/` is still on the table. If not, I won't mention that part.

Reply:
It is on the table (if someone has the time!) but nevertheless I would not mention it. It is very important that documentation always refers to how it is, not how it might be. TODO markers for possible optimizations also have little value as optimizations are driven by profiling, not by looking at TODOs.

Reply:
> refers to how it is, not how it might be

I would only have mentioned it, if it would have been implemented soon-ish, e.g. during FLOSS. But I think that's unlikely so I'll leave it.

> TODO markers for possible optimizations also have little value as optimizations are driven by profiling, not by looking at TODOs.

Finding which optimizations are worth the implementation effort and finding real bottlenecks is determined by profiling. Marking locations where an obvious, but maybe not significant (enough), improvement can be made can be done in advance.

In this case: Not merging and re-splitting a `KeySet` is obviously faster since it's less work, what we don't know is if it actually matters.